File: backend/routes/eligibility.py
"""
routes/eligibility.py — Voter eligibility checker endpoint.

Endpoints
---------
POST /api/eligibility
    Evaluate whether a person is eligible to vote in Indian elections
    based on age and citizenship status.

GET /api/eligibility
    Return the eligibility criteria and official portal link.

Business rules (as per ECI guidelines):
- Minimum voting age: 18 years
- Must be an Indian citizen
- Name must appear on the Electoral Roll (validated separately)
"""

from flask import Blueprint, request

from config import VOTING_AGE
from utils  import error_response, success_response, validate_json_body

eligibility_bp = Blueprint("eligibility", __name__)


# ── Core Logic ─────────────────────────────────────────────────────────────

def evaluate_eligibility(age: int, is_citizen: bool) -> dict:
    """
    Apply ECI eligibility rules and return a structured result.

    Args:
        age:        The voter's age in whole years.
        is_citizen: True if the person is an Indian citizen.

    Returns:
        Dict with keys:
            eligible (bool)             — overall pass/fail
            age_eligible (bool)         — age criterion met
            citizenship_eligible (bool) — citizenship criterion met
            reasons (list[str])         — why criteria were NOT met
            explanation (str)           — human-readable summary
            next_steps (list[str])      — actionable next steps
    """
    reasons = []

    # ── Age check ──────────────────────────────────────────────────────────
    age_ok = age >= VOTING_AGE
    if not age_ok:
        years_left = VOTING_AGE - age
        reasons.append(
            f"You are {age} years old. The minimum voting age in India is {VOTING_AGE}. "
            f"You will be eligible in {years_left} year{'s' if years_left != 1 else ''}."
        )

    # ── Citizenship check ──────────────────────────────────────────────────
    citizenship_ok = is_citizen
    if not citizenship_ok:
        reasons.append(
            "Only Indian citizens can vote in Indian elections. "
            "NRIs with valid Indian passports may be eligible for overseas voting — "
            "check the Election Commission of India (ECI) guidelines."
        )

    eligible = age_ok and citizenship_ok

    # ── Build the result for each of the four eligibility combinations ─────
    if eligible:
        explanation = (
            f"Great news! You are {age} years old and an Indian citizen — "
            "you meet all the basic requirements to vote in Indian elections. "
            "The next step is to check if your name is on the Electoral Roll."
        )
        next_steps = [
            "Check if your name is on the Electoral Roll at voters.eci.gov.in.",
            "If not registered, apply online using Form 6 on the Voter Portal.",
            "Download or collect your Voter ID Card (EPIC) from your local BLO.",
            "On election day, carry a valid photo ID to the polling booth.",
            "Find your polling booth at voterportal.eci.gov.in.",
        ]

    elif not age_ok and citizenship_ok:
        explanation = (
            f"You are {age} years old — you need to be at least {VOTING_AGE} to vote in India. "
            "But it's never too early to learn about the process and get your documents ready!"
        )
        next_steps = [
            "Apply for your Aadhaar Card now if you don't have one.",
            "Learn about the election process at eci.gov.in.",
            "Encourage your eligible family members and friends to vote.",
            f"Set a reminder to register as a voter when you turn {VOTING_AGE}.",
            "Follow India's election news to stay informed.",
        ]

    elif age_ok and not citizenship_ok:
        explanation = (
            "Indian elections are open only to Indian citizens. "
            "If you hold an Indian passport and are an NRI, you may be eligible "
            "for overseas voting under special ECI provisions."
        )
        next_steps = [
            "If you are an NRI with an Indian passport, check Form 6A (Overseas Voter) at eci.gov.in.",
            "Visit eci.gov.in for full citizenship and NRI voter eligibility details.",
            "Contact your nearest Indian Embassy or Consulate for assistance.",
            "If seeking Indian citizenship, refer to the Ministry of Home Affairs website.",
        ]

    else:
        # Both age AND citizenship criteria unmet.
        explanation = (
            "Unfortunately, you do not meet the eligibility criteria for voting in Indian elections "
            "at this time — both minimum age (18) and Indian citizenship are required."
        )
        next_steps = [
            "Learn about India's election process at eci.gov.in.",
            "If you plan to obtain Indian citizenship in the future, refer to MHA guidelines.",
            f"If you are under 18, check back when you turn {VOTING_AGE}.",
            "Encourage eligible citizens around you to participate in democracy.",
        ]

    return {
        "eligible":              eligible,
        "age_eligible":          age_ok,
        "citizenship_eligible":  citizenship_ok,
        "reasons":               reasons,
        "explanation":           explanation,
        "next_steps":            next_steps,
    }


# ── Endpoints ──────────────────────────────────────────────────────────────

@eligibility_bp.route("/eligibility", methods=["POST"])
def check_eligibility():
    """
    Check voter eligibility based on age and citizenship.

    Request JSON:
    {
        "age":         22,
        "citizenship": "indian"   (accepts "indian"/"non-indian"/true/false)
    }

    Response JSON (200):
    {
        "eligible":             true,
        "age_eligible":         true,
        "citizenship_eligible": true,
        "reasons":              [],
        "explanation":          "...",
        "next_steps":           [...]
    }

    Error JSON (400):
    { "error": "..." }
    """
    data = request.get_json(silent=True)

    # ── Validate request body ──────────────────────────────────────────────
    err = validate_json_body(data, "age", "citizenship")
    if err:
        return error_response(err)

    # ── Validate and coerce age ────────────────────────────────────────────
    try:
        age = int(data["age"])
        if not 0 <= age <= 130:
            raise ValueError("Age out of realistic range")
    except (ValueError, TypeError):
        return error_response("'age' must be an integer between 0 and 130")

    # ── Validate and coerce citizenship ───────────────────────────────────
    citizenship_raw = data["citizenship"]
    if isinstance(citizenship_raw, bool):
        is_citizen = citizenship_raw
    elif isinstance(citizenship_raw, str):
        is_citizen = citizenship_raw.strip().lower() in ("indian", "yes", "true", "1")
    else:
        return error_response(
            "'citizenship' must be 'indian', 'non-indian', or a boolean"
        )

    result = evaluate_eligibility(age, is_citizen)
    return success_response(result)


@eligibility_bp.route("/eligibility", methods=["GET"])
def eligibility_info():
    """
    Return the official eligibility criteria and ECI portal link.

    Useful for clients that want to display the criteria without
    performing a check.

    Response JSON (200):
    {
        "criteria": {
            "minimum_age": 18,
            "citizenship": "Indian citizen",
            "enrollment":  "Name must appear on the Electoral Roll"
        },
        "authority": "Election Commission of India (ECI)",
        "portal":    "https://voters.eci.gov.in"
    }
    """
    return success_response({
        "criteria": {
            "minimum_age": VOTING_AGE,
            "citizenship": "Indian citizen",
            "enrollment":  "Name must appear on the Electoral Roll",
        },
        "authority": "Election Commission of India (ECI)",
        "portal":    "https://voters.eci.gov.in",
    })

File: frontend/js/eligibility.js
// ── CivicGuide AI — Eligibility Checker JS ───────────────

const API_URL = "http://127.0.0.1:5000/api/eligibility";

// ─── DOM References ───────────────────────────────────────
const form          = document.getElementById("eligibility-form");
const ageInput      = document.getElementById("age-input");
const ageError      = document.getElementById("age-error");
const checkBtn      = document.getElementById("check-btn");
const spinner       = document.getElementById("spinner");
const btnLabel      = document.getElementById("btn-label");

const resultEmpty   = document.getElementById("result-empty");
const resultCard    = document.getElementById("result-card");
const resultBanner  = document.getElementById("result-banner");
const resultIcon    = document.getElementById("result-icon");
const resultVerdict = document.getElementById("result-verdict");
const criteriaBadges= document.getElementById("criteria-badges");
const resultExpl    = document.getElementById("result-explanation");
const nextStepsList = document.getElementById("next-steps-list");
const resultActions = document.getElementById("result-actions");

// ─── Helpers ──────────────────────────────────────────────
function setLoading(on) {
  checkBtn.disabled     = on;
  spinner.style.display = on ? "block" : "none";
  btnLabel.textContent  = on ? "Checking…" : "Check Eligibility →";
}

function showAgeError(msg) {
  ageError.textContent  = msg;
  ageError.style.display = "block";
  ageInput.style.borderColor = "rgba(248,113,113,0.6)";
}

function clearAgeError() {
  ageError.style.display = "none";
  ageInput.style.borderColor = "";
}

function getCitizenshipValue() {
  return document.querySelector('input[name="citizenship"]:checked')?.value ?? "indian";
}

// ─── Render Result ────────────────────────────────────────
function renderResult(data) {
  const { eligible, age_eligible, citizenship_eligible,
          explanation, next_steps } = data;

  // Banner class
  resultBanner.className = `result-banner ${eligible ? "eligible-banner" : "ineligible-banner"}`;

  // Icon & verdict
  if (eligible) {
    resultIcon.textContent    = "✅";
    resultVerdict.textContent = "Eligible to Vote!";
  } else if (!age_eligible && citizenship_eligible) {
    resultIcon.textContent    = "⏳";
    resultVerdict.textContent = "Not Yet Eligible (Age)";
  } else if (age_eligible && !citizenship_eligible) {
    resultIcon.textContent    = "🌍";
    resultVerdict.textContent = "Not Eligible (Citizenship)";
  } else {
    resultIcon.textContent    = "❌";
    resultVerdict.textContent = "Not Eligible";
  }

  // Criteria badges
  criteriaBadges.innerHTML = `
    <span class="badge ${age_eligible ? "pass" : "fail"}">
      ${age_eligible ? "✔" : "✘"} Age
    </span>
    <span class="badge ${citizenship_eligible ? "pass" : "fail"}">
      ${citizenship_eligible ? "✔" : "✘"} Citizenship
    </span>
  `;

  // Explanation
  resultExpl.textContent = explanation;

  // Next steps
  nextStepsList.innerHTML = next_steps
    .map((step, i) => `
      <li>
        <span class="step-num">${i + 1}</span>
        <span>${escapeHtml(step)}</span>
      </li>`)
    .join("");

  // Action buttons
  if (eligible) {
    resultActions.innerHTML = `
      <a href="https://voters.eci.gov.in" target="_blank" rel="noopener" class="btn-solid">
        Register to Vote →
      </a>
      <a href="chat.html" class="btn-outline">Ask the AI →</a>
    `;
  } else {
    resultActions.innerHTML = `
      <a href="chat.html" class="btn-solid">Ask CivicGuide AI →</a>
      <a href="https://eci.gov.in" target="_blank" rel="noopener" class="btn-outline">
        Visit ECI →
      </a>
    `;
  }

  // Show result, hide empty
  resultEmpty.style.display = "none";
  resultCard.style.display  = "block";
  resultCard.style.animation = "none";
  // Trigger re-animation
  requestAnimationFrame(() => {
    resultCard.style.animation = "";
  });

  // Scroll result into view on mobile
  if (window.innerWidth < 768) {
    resultCard.scrollIntoView({ behavior: "smooth", block: "start" });
  }
}

function escapeHtml(str) {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

// ─── Form Submit ──────────────────────────────────────────
form.addEventListener("submit", async (e) => {
  e.preventDefault();
  clearAgeError();

  // Validate age
  const ageRaw = ageInput.value.trim();
  if (!ageRaw) { showAgeError("Please enter your age."); return; }

  const age = parseInt(ageRaw, 10);
  if (isNaN(age) || age < 0 || age > 130) {
    showAgeError("Please enter a valid age between 0 and 130.");
    return;
  }

  const citizenship = getCitizenshipValue();

  setLoading(true);

  try {
    const res = await fetch(API_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ age, citizenship })
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.error || `Server error ${res.status}`);
    }

    const data = await res.json();
    renderResult(data);

  } catch (err) {
    console.error("[CivicGuide AI] Eligibility error:", err);
    // Graceful offline fallback — run logic client-side
    const fallback = clientFallback(age, citizenship === "indian");
    renderResult(fallback);
  } finally {
    setLoading(false);
  }
});

// ─── Client-side Fallback (offline) ──────────────────────
function clientFallback(age, isCitizen) {
  const ageOk  = age >= 18;
  const eligible = ageOk && isCitizen;

  let explanation, next_steps;

  if (eligible) {
    explanation = `You are ${age} years old and an Indian citizen — you are eligible to vote in Indian elections!`;
    next_steps  = [
      "Check your name on the Electoral Roll at voters.eci.gov.in.",
      "If not registered, apply using Form 6 on the Voter Portal.",
      "Collect your Voter ID Card (EPIC).",
      "Find your polling booth on election day."
    ];
  } else if (!ageOk && isCitizen) {
    explanation = `You are ${age} years old. You must be at least 18 to vote. You will be eligible in ${18 - age} year(s).`;
    next_steps  = [
      "Get your Aadhaar Card if you don't have one.",
      "Learn about elections at eci.gov.in.",
      `Register as a voter when you turn 18.`,
      "Encourage eligible family members to vote."
    ];
  } else if (ageOk && !isCitizen) {
    explanation = "Only Indian citizens can vote in Indian elections. NRIs with Indian passports may have special provisions.";
    next_steps  = [
      "Check NRI overseas voter registration (Form 6A) at eci.gov.in.",
      "Contact your nearest Indian Embassy or Consulate.",
      "Visit eci.gov.in for full citizenship eligibility details."
    ];
  } else {
    explanation = "You do not meet the eligibility criteria — both age (18+) and Indian citizenship are required.";
    next_steps  = [
      "Learn about the election process at eci.gov.in.",
      "Check back when you meet the age and citizenship requirements."
    ];
  }

  return {
    eligible,
    age_eligible: ageOk,
    citizenship_eligible: isCitizen,
    reasons: [],
    explanation,
    next_steps
  };
}

// ─── Clear error on input change ─────────────────────────
ageInput.addEventListener("input", clearAgeError);
